But what happens if one of the async components fails to retrieve its data? Are we stuck with the loading indicator?

No, the fallback content disappears when all components are resolved, successfully or not. So we end up with just the component in success showing up.

We can catch the potential errors though, by using the onErrorCaptured lifecycle hook. onErrorCaptured can be used in a component to capture any error from a descendent component.

You can do whatever you need to handle the error, and return true (or donâ€™t return) to let it propagate up the chain of components, or return false to stop its propagation.